tls/s2n_ktls_io.c

```diff
@@ -65,3 +83,92 @@ static ssize_t s2n_ktls_default_sendmsg(void *io_context, const struct msghdr *m
 
     return sendmsg(fd, msg, 0);
 }
+
+S2N_RESULT s2n_ktls_set_control_data(struct msghdr *msg, char *buf, size_t buf_size,
+        int cmsg_type, uint8_t record_type)
+{
+    RESULT_ENSURE_REF(msg);
+    RESULT_ENSURE_REF(buf);
+
+    /*
+     * https://man7.org/linux/man-pages/man3/cmsg.3.html
+     * To create ancillary data, first initialize the msg_controllen
+     * member of the msghdr with the length of the control message
+     * buffer.
+     */
+    msg->msg_control = buf;
+    msg->msg_controllen = buf_size;
+
+    /*
+     * https://man7.org/linux/man-pages/man3/cmsg.3.html
+     * Use CMSG_FIRSTHDR() on the msghdr to get the first
+     * control message and CMSG_NXTHDR() to get all subsequent ones.
+     */
+    struct cmsghdr *hdr = CMSG_FIRSTHDR(msg);
+    RESULT_ENSURE_REF(hdr);
+
+    /*
+     * https://man7.org/linux/man-pages/man3/cmsg.3.html
+     * In each control message, initialize cmsg_len (with CMSG_LEN()), the
+     * other cmsghdr header fields, and the data portion using
+     * CMSG_DATA().
+     */
+    hdr->cmsg_len = CMSG_LEN(S2N_KTLS_RECORD_TYPE_SIZE);
+    hdr->cmsg_level = S2N_SOL_TLS;
+    hdr->cmsg_type = cmsg_type;
+    *CMSG_DATA(hdr) = record_type;
+
+    /*
+     * https://man7.org/linux/man-pages/man3/cmsg.3.html
+     * Finally, the msg_controllen field of the msghdr
+     * should be set to the sum of the CMSG_SPACE() of the length of all
+     * control messages in the buffer
+     */
+    RESULT_ENSURE_GTE(msg->msg_controllen, CMSG_SPACE(S2N_KTLS_RECORD_TYPE_SIZE));
+    msg->msg_controllen = CMSG_SPACE(S2N_KTLS_RECORD_TYPE_SIZE);
+
+    return S2N_RESULT_OK;
+}
+
+/* Expect to receive a single cmsghdr containing the TLS record_type.
+ *
+ * s2n-tls allocates enough space to receive a single cmsghdr. Since this is
+ * used to get the record_type when receiving over kTLS (enabled via
+ * `s2n_connection_ktls_enable_recv`), the application should not configure
+ * the socket to receive additional control messages. In the event s2n-tls
+ * can not retrieve the record_type, it is safer to drop the record.
```

Comment:
"drop the record" isn't a thing in TLS like it would be in QUIC. We kill the connection on an error.